qip-0014.md

```diff
@@ -0,0 +1,170 @@
+```
+ QIP: 14
+ Layer: EVM
+ Title: CREATE + CREATE2 opcode
+ Author: alanorwick <alan+qip@quai.org>
+ Comments-Summary: No comments yet.
+ Comments-URI: https://github.com/quainetwork/qips/wiki/Comments:QIP-0014
+ Status: Draft
+ Type: Informational
+ Created: 2024-07-31
+ License: BSD-2-Clause
+```
+
+
+## Abstract
+
+This QIP proposes modifications to the CREATE opcode in Quai Network to ensure shard-specific address generation, while maintaining CREATE2 functionality as is. The proposal aims to adapt Ethereum's address creation mechanism to Quai's sharded architecture.
+
+## Motivation
+
+Quai Network's sharded architecture requires addresses to be shard-specific. The current Ethereum-style CREATE opcode does not guarantee this, potentially leading to cross-shard conflicts and inefficiencies. This proposal seeks to address this issue while maintaining as much compatibility with Ethereum as possible.
+
+## Specification
+
+### CREATE Opcode Modification
+
+1. The CREATE opcode will generate an address using the following method:
+   
+   a. Split the bytecode_data into two parts:
+      - `hashCode`: All but the last 4 bytes of bytecode_data
+      - `potentialSalt`: The last 4 bytes of bytecode_data
+   
+   b. Calculate the initial address:
+
+      ```address = keccak256(([sender_address, sender_nonce, keccak256(hashCode), potentialSalt])[12:]```
+   
+   c. If the resulting address is valid for the current shard:
+      - The full bytecode_data (including the last 4 bytes) is used as the contract bytecode
+      - The `potentialSalt` is considered part of the actual bytecode
+   
+   d. If the resulting address is not valid for the current shard:
+      - Proceed to step 2 for grinding
+
+2. If the resulting address is not valid for the current shard:
+The gas cost per iteration of bytecode modification and checking is the `KeccackGasCost` of the size of `concat(sender_address, sender_nonce, hashCodeHash, salt)`.
+  
+   a. Initialize a "grind nonce" to 0.
+  
+   b. Append 4 bytes representing the grind nonce to the original bytecode_data:
+      ```
+      extended_bytecode = bytecode_data + [4]byte(grind_nonce)
+   ```
+   c. Split the extended_bytecode:
+      - `hashCode`: All but the last 4 bytes of extended_bytecode
+      - `salt`: The last 4 bytes of extended_bytecode
+   
+   d. Generate a new address using:
+      `address = keccak256([sender_address, sender_nonce, keccak256(hashCode), salt])[12:]`
+   
+   e. If the address is still not valid, increment the grind nonce and repeat steps b, c, and d until a valid shard-specific address is found, until 1,000 tries has been exceeded, or the gas limit has been exceed.
+   d. Exceeding 1,000 tries without finding a valid shard address will result in a error, thus failing the contract deployment.
```

Comment:
Why arbitrarily cap at 1000 iterations? This is the reason the gas limit exists. If a user is willing to pay the gas for 1001 iterations, then they should get all 1001 iterations.

Reply:
I agree with this approach. 1,000 tries was from the prior implementation.